src/libcore/num/f32.rs

```diff
@@ -474,4 +475,85 @@ impl f32 {
+    /// defines an ordering between all distinct bit patterns:
+    ///
+    /// - For normal numbers, the ordering matches the comparison operators.
+    /// - Zeros are between the positive and negative normal numbers, with
+    ///   the positive zero greater than the negative zero.
+    /// - Positive infinity is greater than all normal numbers; negative
+    ///   infinity is less than all normal numbers.
+    /// - Positive NaNs are greater than all other values, ordered amongst
+    ///   themselves by their payloads; Negative NaNs are less than all other
+    ///   values, ordered amongst in reverse order of their payloads.
+    ///
+    /// # Examples
+    ///
+    /// Normal numbers:
+    /// ```
+    /// #![feature(float_total_cmp)]
+    /// use std::cmp::Ordering::*;
+    ///
+    /// assert_eq!(f32::total_cmp(1.0, 2.0), Less);
+    /// assert_eq!(f32::total_cmp(1.0, 1.0), Equal);
+    /// assert_eq!(f32::total_cmp(2.0, 1.0), Greater);
+    /// assert_eq!(f32::total_cmp(-1.0, 1.0), Less);
+    /// assert_eq!(f32::total_cmp(1.0, -1.0), Greater);
+    /// assert_eq!(f32::total_cmp(-1.0, -2.0), Greater);
+    /// assert_eq!(f32::total_cmp(-1.0, -1.0), Equal);
+    /// assert_eq!(f32::total_cmp(-2.0, -1.0), Less);
+    /// ```
+    ///
+    /// Zeros, infinities, and NaNs:
+    /// ```
+    /// #![feature(float_total_cmp)]
+    /// use std::cmp::Ordering::*;
+    ///
+    /// assert_eq!(f32::partial_cmp(&0.0, &-0.0), Some(Equal));
+    /// assert_eq!(f32::total_cmp(-0.0, 0.0), Less);
+    /// assert_eq!(f32::total_cmp(0.0, -0.0), Greater);
+    ///
+    /// assert_eq!(f32::partial_cmp(&std::f32::NAN, &std::f32::NAN), None);
+    /// assert_eq!(f32::total_cmp(-std::f32::NAN, std::f32::NAN), Less);
+    /// assert_eq!(f32::total_cmp(std::f32::NAN, std::f32::NAN), Equal);
+    /// assert_eq!(f32::total_cmp(std::f32::NAN, -std::f32::NAN), Greater);
+    /// assert_eq!(f32::total_cmp(std::f32::NAN, std::f32::INFINITY), Greater);
+    /// assert_eq!(f32::total_cmp(-std::f32::NAN, -std::f32::INFINITY), Less);
+    /// ```
+    ///
+    /// Using this to sort floats:
+    /// ```
+    /// #![feature(float_total_cmp)]
+    ///
+    /// let mut a = [
+    ///     1.0,
+    ///     -1.0,
+    ///     0.0,
+    ///     -0.0,
+    ///     std::f32::NAN,
+    ///     -std::f32::NAN,
+    ///     std::f32::INFINITY,
+    ///     std::f32::NEG_INFINITY,
+    /// ];
+    /// a.sort_by(|&x, &y| f32::total_cmp(x, y));
+    /// assert_eq!(
+    ///     format!("{:?}", a),
+    ///     "[NaN, -inf, -1.0, -0.0, 0.0, 1.0, inf, NaN]"
+    /// );
+    /// ```
+    #[unstable(feature = "float_total_cmp", issue = "88888888")]
+    #[inline]
+    pub fn total_cmp(self, other: f32) -> cmp::Ordering {
+        self.total_order_key().cmp(&other.total_order_key())
+    }
+
+    fn total_order_key(self) -> i32 {
+        let x = self.to_bits() as i32;
+        // Flip the bottom 31 bits if the high bit is set
+        x ^ ((x >> 31) as u32 >> 1) as i32
```

Comment:
I don't have a proof for this, but it's equivalent to the one in https://github.com/rust-lang/rfcs/issues/1249#issuecomment-368639944.  I suspect the predicate was chosen specifically to allow such a simple implementation, because of things like "the same floating-point datum" being ordered by exponent first.

Reply:
I have the exact same question for @jpetkau then 😉 I agree with your suspicion (tho note that "same floating point datum" with different exponents only happens in decimal floating point) but I'd rather have it confirmed.

Reply:
How about something like:

"With their bits interpreted as signed integers, positive IEEE754 floats are ordered as:

    +0 < (positive finite floats in order) < +inf < +NANs.

With the reverse order for floats with the sign bit set.

So if we reverse the order of the negative cases, by xoring the sign bit with the other bits, we get the order:

    -NANs < -inf < (negative finite floats in order) < -0 < +0 < (positive finite floats in order) < +inf < +NANs.

IEEE754.2008 calls this the 'totalOrder' predicate.

(It also specifies an interpretation of the signaling bit which implies -QNANs < -SNANs < +SNANs < +QNANs, but if we happen to be on an older MIPS architecture with the opposite interpretation of the signaling bit, this won't be true.)

Reply:
I wonder whether the behavior of `.to_bits()` (which does a `transmute` internally) is endian dependent, and whether that matters for ordering purposes (e.g. the sign bit in a big-endian target being at a different offset).

Reply:
Endianness doesn't matter, because an f32 and the i32 will still have the same endianness as each other. It's still a 32-bit word, this is just interpreting the bits as one type or another.

It's no different than any case where you might encode a value into an int (e.g. suppose you wanted to compactly store three values, of size 1, 8, and 23 bits, inside a 32-bit int. You could make a `struct Foo(u32)`, with `fn make_foo(u32 a, u32 b, u32 c) -> Foo { Foo((a & 1)<<31 | (b&0xff)<<23 | (c&0x07ff)) }`.

This is literally all that's inside an IEEE float. The transmute is just because we can't get at the bits with `foo.0`.

Reply:
It's true that this reinterprets the whole 32- or 64-bit unit, but it's possible that an architecture [uses different endianness for floats and ints](https://en.wikipedia.org/wiki/Endianness#Floating_point), then the bytes will appear reversed when stored to memory as a float and loaded back as an integer. However, I don't think we currently support such architectures, and if we wanted to, we'd have to update more code than just this method (in particular `to_bits`, which would automatically make *this* code work correctly too).

Reply:
I dispute that it's possible that an architecture uses different endianness for floats and ints. That would be deeply insane and break a lot more than this code.

Reply:
What do you mean you dispute that it's possible? There's real physical processors (old ARMs) doing so. We don't support them, and I would be open to concluding we don't ever want to support them (just like e.g. architectures with non-octet memory granularity &ndash; also breaks *tons* of code, but definitely a real thing!), but they exist.

Reply:
Sorry, I was just being flippant there. Such architectures do exist. But they would break far more than just this code; they'd break everything that relied on the bit pattern of floats, which is, well, *everything*. Every serialization library, formatter, parser, implementation of math libraries, you name it.

Reply:
> Endianness doesn't matter, because an f32 and the i32 will still have the same endianness as each other.

Thanks for the answer, I just wasn't sure if this had to be the case, but it seems that for Rust this will need to be the case.